### PR TITLE
Add gather operation

### PR DESCRIPTION
**Pull request by yanndupis**
Hey @mortendahl ,

This PR add the operation [tf.gather](https://www.tensorflow.org/api_docs/python/tf/gather).

Would you mind review it?

Thanks

### CODE REVIEW

File: tf_encrypted/tensor/int100.py [hunk 1]
```diff
@@ -568,6 +568,10 @@ def strided_slice(self, args, kwargs):
             backing = [tf.strided_slice(xi, *args, **kwargs) for xi in self.backing]
             return DenseTensor(backing)
 
+        def gather(self, indices: list, axis: int = 0):
+            backings = [tf.gather(xi, indices, axis=axis) for xi in self.backing]
+            return [DenseTensor(backing) for backing in backings]
```
Comment (mortendahl) [inline tf_encrypted/tensor/int100.py]: I'm not sure why you want to return a list here? Shouldn't it just be a single tensor?

File: tf_encrypted/protocol/pond.py [hunk 2]
```diff
@@ -3215,6 +3221,76 @@ def _strided_slice_masked(prot, x: PondMaskedTensor, args: Any, kwargs: Any):
...
 
 
+#
+# gather helpers
+#
+
+
+def _gather_public(
+    prot: Pond, x: PondPublicTensor, indices: list, axis: int = 0
+) -> List[PondPublicTensor]:
```
Comment (mortendahl) [inline tf_encrypted/protocol/pond.py]: you're returning a single `PondPublicTensor`, not a list

File: tf_encrypted/protocol/pond.py [hunk 3]
```diff
@@ -3215,6 +3221,76 @@ def _strided_slice_masked(prot, x: PondMaskedTensor, args: Any, kwargs: Any):
...
+
+    x_on_0, x_on_1 = x.unwrapped
+
+    with tf.name_scope("gather"):
+
+        with tf.device(prot.server_0.device_name):
+            y0_g = x_on_0.gather(indices, axis=axis)
```
Comment (mortendahl) [inline tf_encrypted/protocol/pond.py]: consider renaming to `y_on_0_g` for consistency (public value vs share)

File: tf_encrypted/protocol/pond.py [hunk 4]
```diff
@@ -3215,6 +3221,76 @@ def _strided_slice_masked(prot, x: PondMaskedTensor, args: Any, kwargs: Any):
...
+
+        with tf.device(prot.server_1.device_name):
+            y1_g = x_on_1.gather(indices, axis=axis)
+
+        return PondPublicTensor(prot, y0_g, y1_g, x.is_scaled)
+
+
+def _gather_private(
+    prot: Pond, x: PondPrivateTensor, indices: list, axis: int = 0
+) -> List[PondPrivateTensor]:
```
Comment (mortendahl) [inline tf_encrypted/protocol/pond.py]: return type, see other comment

File: tf_encrypted/protocol/pond.py [hunk 5]
```diff
@@ -3215,6 +3221,76 @@ def _strided_slice_masked(prot, x: PondMaskedTensor, args: Any, kwargs: Any):
...
+            y0_g = x0.gather(indices, axis=axis)
+
+        with tf.device(prot.server_1.device_name):
+            y1_g = x1.gather(indices, axis=axis)
+
+        return PondPrivateTensor(prot, y0_g, y1_g, x.is_scaled)
+
+
+def _gather_masked(
+    prot: Pond, x: PondMaskedTensor, shape: List[int]
```
Comment (mortendahl) [inline tf_encrypted/protocol/pond.py]: signature is different from other gather methods, copy&paste error?

File: tf_encrypted/protocol/pond.py [hunk 6]
```diff
@@ -3215,6 +3221,76 @@ def _strided_slice_masked(prot, x: PondMaskedTensor, args: Any, kwargs: Any):
...
+            a0_g = a0.reshape(shape)
+            alpha_on_0_g = alpha_on_0.gather(shape)
+
+        with tf.device(prot.server_1.device_name):
+            a1_g = a1.reshape(shape)
+            alpha_on_1_g = alpha_on_1.gather(shape)
+
+        return PondMaskedTensor(
+            prot,
+            prot.reshape(x.unmasked, shape),
```
Comment (mortendahl) [inline tf_encrypted/protocol/pond.py]: looks like copy&paste error; there are a few more in this function

File: tf_encrypted/protocol/pond.py [hunk 7]
```diff
@@ -907,6 +907,12 @@ def strided_slice(self, x: "PondTensor", *args: Any, **kwargs: Any):
 
         return x_sliced
 
+    @memoize
+    def gather(
+        self, x: "PondTensor", indices: list, axis: int = 0
+    ) -> List["PondTensor"]:
```
Comment (mortendahl) [inline tf_encrypted/protocol/pond.py]: not sure it should return a list per the TF api

File: tf_encrypted/tensor/int100.py (hunk 1 above)
Comment (yanndupis) [reply tf_encrypted/tensor/int100.py]: ah good catch, I had fixe it in `native` but not in int100. thanks!

File: tf_encrypted/protocol/pond.py (hunk 3 above)
Comment (yanndupis) [reply tf_encrypted/protocol/pond.py]: good point!

**Comment by yanndupis** [comment]
Thanks @mortendahl for your review. This pr had definitely couple issues. It should be fixed now.

**Pull request merged**